File: projects/engineering.purdue.edu/kak/programmingwithobjects/pwocode/chap11/Static1.java
// This code example is from the following source:
//
// Book Title:  Programming with Objects, A Comparative Presentation
//              of Object-Oriented Programming with C++ and Java
//
// Book Author:   Avinash Kak
//
// Chapter:     Chapter 11  Classes, The Rest Of The Story
//
// Section:     Section 11.4  Static Members In Java
//



//Static1.java

class X {
    public int m;                                          
    public static int n = 100;                             
    public static final int p = 50;                               //(A)
    public static final int q = init();                           //(B)

    public X( int mm ) { m = mm; }                         

    public static int init() { return 200; }
}

class Test {
    public static void main(String[] args)
    {
        System.out.println( X.n );        // 100                  //(C)
        System.out.println( X.q );        // 200 
    
        X xobj_1 = new X( 20 );                                   //(D)
        System.out.println(xobj_1.m + " " + xobj_1.n);   //20 100 
                                                         
        X xobj_2 = new X( 40 );                         
        System.out.println(xobj_2.m + " " + xobj_2.n);   //40 100  
                                                         
        X.n = 1000;                                               //(E) 
  
        System.out.println(xobj_1.m + " " + xobj_1.n);   //20 1000   
                                                        
        System.out.println(xobj_2.m + " " + xobj_2.n);   //40 1000   
                                                        
    }
}

File: projects/engineering.purdue.edu/kak/programmingwithobjects/pwocode/chap11/Static1.cc
// This code example is from the following source:
//
// Book Title:  Programming with Objects, A Comparative Presentation
//              of Object-Oriented Programming with C++ and Java
//
// Book Author:   Avinash Kak
//
// Chapter:     Chapter 11  Classes, The Rest Of The Story
//
// Section:     Section 11.3  Static Members In C++
//



//Static1.cc

#include <iostream>
using namespace std;

class X {
public:
    int m;                                          
    static int n;                                   
    static const int p;                             
    static const int q = 51;                        
    X( int mm ) { m = mm; }                         
};

int X::n = 100;                                                   //(A)

const int X::p = 101;                                             //(B)

int main()
{
    cout << X::n << endl;          // 100                         //(C)
  
    X xobj_1( 20 );                                   
    cout << xobj_1.m << " " << xobj_1.n << endl;  // 20 100       //(D)
  
    X xobj_2( 40 );                                   
    cout << xobj_2.m << " " << xobj_2.n << endl;  // 40 100

    X::n = 1000;                                                  //(E)
  
    cout << xobj_1.m << " " << xobj_1.n << endl;  // 20 1000
    cout << xobj_2.m << " " << xobj_2.n << endl;  // 40 1000

    return 0;
}
